File: backend/src/marketData/backfill.ts
import { prisma } from "../db/client.js";
import { childLogger } from "../core/logger.js";
import { ACTIVE_INSTRUMENTS, DEFAULT_INSTRUMENTS, type InstrumentSpec } from "./instruments.js";
import { fetchYahooChart, type YahooBar } from "./yahooClient.js";

const logger = childLogger("backfill");

// Yahoo's free chart endpoint does not serve interval=1m for CME futures
// continuous contracts (=F symbols) -- only for equities. 5-minute is the
// finest granularity actually available for these symbols, going back ~60
// days; `bars_1m` (the table name is a holdover from the original spec)
// stores this 5-minute data for futures.
const INTRADAY_INTERVAL = "5m";
const MAX_INTRADAY_LOOKBACK_DAYS = 60;

export async function ensureInstrumentsSeeded(): Promise<void> {
  for (const spec of DEFAULT_INSTRUMENTS) {
    await prisma.instrument.upsert({
      where: { symbol: spec.symbol },
      // tickSize/pointValue must stay in sync with DEFAULT_INSTRUMENTS on every
      // boot, not just at first creation -- engine/accounting.ts reads
      // pointValue from this table (unlike the rest of the engine, which reads
      // getInstrument() directly), so a stale DB row after a contract-spec
      // change (e.g. switching from full-size to micro contracts) would
      // silently miscalculate live unrealized P&L.
      update: {
        dataSymbol: spec.dataSymbol,
        exchange: spec.exchange,
        tickSize: spec.tickSize.toString(),
        pointValue: spec.pointValue.toString(),
      },
      create: {
        symbol: spec.symbol,
        dataSymbol: spec.dataSymbol,
        exchange: spec.exchange,
        tickSize: spec.tickSize.toString(),
        pointValue: spec.pointValue.toString(),
      },
    });
  }
}

export async function backfillDaily(spec: InstrumentSpec, days: number): Promise<number> {
  const range = `${Math.max(days, 30)}d`;
  const bars = await fetchYahooChart(spec.dataSymbol, range, "1d");
  if (bars.length === 0) {
    logger.warn({ symbol: spec.symbol }, "daily_backfill_empty");
    return 0;
  }

  // Historical daily bars for past dates don't change -- bulk-insert and
  // skip rows that already exist instead of one network round-trip per row
  // (an N-row loop of individual upserts against a remote DB was the actual
  // bottleneck behind what first looked like Yahoo rate-limiting).
  const rows = bars.map((bar: YahooBar) => {
    const date = new Date(Date.UTC(bar.time.getUTCFullYear(), bar.time.getUTCMonth(), bar.time.getUTCDate()));
    return {
      date, symbol: spec.symbol,
      open: bar.open.toString(), high: bar.high.toString(), low: bar.low.toString(), close: bar.close.toString(), volume: bar.volume.toString(),
    };
  });
  await prisma.dailyBar.createMany({ data: rows, skipDuplicates: true });

  logger.info({ symbol: spec.symbol, rows: bars.length }, "daily_backfill_done");
  return bars.length;
}

export async function backfillRecentIntraday(spec: InstrumentSpec): Promise<number> {
  const bars = await fetchYahooChart(spec.dataSymbol, `${MAX_INTRADAY_LOOKBACK_DAYS}d`, INTRADAY_INTERVAL);
  if (bars.length === 0) {
    logger.warn({ symbol: spec.symbol }, "intraday_backfill_empty");
    return 0;
  }

  const rows = bars.map((bar: YahooBar) => ({
    time: bar.time, symbol: spec.symbol,
    open: bar.open.toString(), high: bar.high.toString(), low: bar.low.toString(), close: bar.close.toString(), volume: bar.volume.toString(),
  }));
  await prisma.bar.createMany({ data: rows, skipDuplicates: true });

  logger.info({ symbol: spec.symbol, rows: bars.length }, "intraday_backfill_done");
  return bars.length;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function runFullBackfill(days = 365): Promise<void> {
  await ensureInstrumentsSeeded();
  // Only the actively-traded symbols -- see ACTIVE_INSTRUMENTS's comment.
  // CL/GC's existing historical data is untouched, just never re-fetched.
  for (const spec of ACTIVE_INSTRUMENTS) {
    await backfillDaily(spec, days);
    await backfillRecentIntraday(spec);
    // A little breathing room between symbols regardless -- cheap insurance
    // against bursting Yahoo's unauthenticated endpoint.
    await sleep(1000);
  }
}

File: backend/app/market_data/backfill.py
"""Historical data ingestion from the free Yahoo Finance vendor.

Yahoo's public data has real constraints that shape this module's design:
- 1-minute bars are only available for the trailing ~7 days.
- Daily bars are available for the instrument's full history.
So: `backfill_daily()` pulls 1+ year of daily bars (the honest way to satisfy
"a year of historical price data" for free) into `bars_daily`, and
`backfill_recent_intraday()` seeds `bars_1m` with whatever 1-minute history
Yahoo will give us so the live pipeline has continuity when it starts polling.
"""
from __future__ import annotations

import asyncio
import datetime as dt
from decimal import Decimal

import pandas as pd
from sqlalchemy import select
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.logging import get_logger
from app.db.models import Bar, DailyBar, Instrument
from app.market_data.instruments import DEFAULT_INSTRUMENTS, InstrumentSpec

logger = get_logger(__name__)

MAX_INTRADAY_LOOKBACK_DAYS = 7


async def ensure_instruments_seeded(session: AsyncSession) -> None:
    existing = {row.symbol for row in (await session.execute(select(Instrument))).scalars().all()}
    for spec in DEFAULT_INSTRUMENTS:
        if spec.symbol in existing:
            continue
        session.add(
            Instrument(
                symbol=spec.symbol,
                data_symbol=spec.data_symbol,
                exchange=spec.exchange,
                tick_size=spec.tick_size,
                point_value=spec.point_value,
            )
        )
    await session.commit()


def _download_sync(data_symbol: str, period: str, interval: str) -> pd.DataFrame:
    import yfinance as yf

    df = yf.Ticker(data_symbol).history(period=period, interval=interval, auto_adjust=False)
    return df


async def backfill_daily(session: AsyncSession, spec: InstrumentSpec, days: int) -> int:
    period = f"{max(days, 30)}d"
    df = await asyncio.to_thread(_download_sync, spec.data_symbol, period, "1d")
    if df.empty:
        logger.warning("backfill.daily.empty", symbol=spec.symbol)
        return 0

    rows = [
        {
            "date": idx.date(),
            "symbol": spec.symbol,
            "open": Decimal(str(round(float(row["Open"]), 8))),
            "high": Decimal(str(round(float(row["High"]), 8))),
            "low": Decimal(str(round(float(row["Low"]), 8))),
            "close": Decimal(str(round(float(row["Close"]), 8))),
            "volume": Decimal(str(float(row["Volume"] or 0))),
        }
        for idx, row in df.iterrows()
    ]

    stmt = pg_insert(DailyBar).values(rows)
    stmt = stmt.on_conflict_do_update(
        index_elements=[DailyBar.date, DailyBar.symbol],
        set_={"open": stmt.excluded.open, "high": stmt.excluded.high, "low": stmt.excluded.low,
              "close": stmt.excluded.close, "volume": stmt.excluded.volume},
    )
    await session.execute(stmt)
    await session.commit()
    logger.info("backfill.daily.done", symbol=spec.symbol, rows=len(rows))
    return len(rows)


async def backfill_recent_intraday(session: AsyncSession, spec: InstrumentSpec) -> int:
    df = await asyncio.to_thread(_download_sync, spec.data_symbol, f"{MAX_INTRADAY_LOOKBACK_DAYS}d", "1m")
    if df.empty:
        logger.warning("backfill.intraday.empty", symbol=spec.symbol)
        return 0

    rows = [
        {
            "time": idx.to_pydatetime(),
            "symbol": spec.symbol,
            "open": Decimal(str(round(float(row["Open"]), 8))),
            "high": Decimal(str(round(float(row["High"]), 8))),
            "low": Decimal(str(round(float(row["Low"]), 8))),
            "close": Decimal(str(round(float(row["Close"]), 8))),
            "volume": Decimal(str(float(row["Volume"] or 0))),
        }
        for idx, row in df.iterrows()
    ]

    stmt = pg_insert(Bar).values(rows)
    stmt = stmt.on_conflict_do_update(
        index_elements=[Bar.time, Bar.symbol],
        set_={"open": stmt.excluded.open, "high": stmt.excluded.high, "low": stmt.excluded.low,
              "close": stmt.excluded.close, "volume": stmt.excluded.volume},
    )
    await session.execute(stmt)
    await session.commit()
    logger.info("backfill.intraday.done", symbol=spec.symbol, rows=len(rows))
    return len(rows)


async def run_full_backfill(session: AsyncSession, days: int = 365) -> None:
    await ensure_instruments_seeded(session)
    for spec in DEFAULT_INSTRUMENTS:
        await backfill_daily(session, spec, days)
        await backfill_recent_intraday(session, spec)
